### ISSUE TITLE
Rebuilding or serving library project

### ISSUE THREAD

**Issue by Nodarii** [angular/angular-cli#10643]
### Feature Request
Some interface that would rebuild library project that is generated with `ng g library niceLib`.
Of Course the best option would be `ng serve niceLib` or include it to the application serve. Or at least another build that allow to add `--watch` parameter like it does for for application. Here what I mean:
* for application we are allowed to run `ng build --watch`
* for library we have `ng build niceLib`
* but we can't use `ng build niceLib --watch`, as it throws `Unknown option: '--watch'`

**Comment by Tyler-V** [comment]
The new angular.json specifies a `'defaultProject'` of which you generate and then are able to use the generate library command which creates the subsequent projects/ and updates the angular.json.

What I'm doing is using my `'defaultProject'` as the demo app and then `import { ProjectModule} from 'projects/name/src/public_api'`

Any file changes inside `'defaultProject'` or libraries `projects/` reloads the development server

**Comment by filipesilva** [comment]
Heya, this feature request is something we are aware of and talk about in https://github.com/angular/angular-cli/wiki/stories-create-library#why-do-i-need-to-build-the-library-everytime-i-make-changes.


> In the future we want to add watch support to building libraries so it is faster to see changes.
>
> We are also planning to add internal dependency support to Angular CLI. This means that Angular CLI would know your app depends on your library, and automatically rebuilds the library when the app needs it.

**Comment by beeman** [comment]
As a workaround you can use `nodemon` to rebuild the libraries:

`npx nodemon -w ./projects/ui -e ts --exec 'ng build ui'`

It's on the slow side but does get the job done. One downside is that you need to run a separate process for each of the libs you want to rebuild.

**Comment by hwangzhiming** [comment]
try `File Watchers` if you use `WebStorm`.

**Comment by giggio** [comment]
The proposal from @Tyler-V is by far the only acceptable at this moment. No developer can take the productivity hit that manually rebuilding the library causes. Even with an automated watcher the build time increases substantially. Until there is a fix I, differently from what the ng team proposes, will err on the side of productivity instead of caution.

**Comment by Tyler-V** [comment]
@giggio, 

@filipesilva  and @alan-agius4 worked on some changes this morning and I'm currently able to serve my library from my demo project as suggested and build all in --aot!

I believe changes were committed and merged on ng-packagr to support this.

https://github.com/dherges/ng-packagr/pull/862

**Comment by giggio** [comment]
@Tyler-V That is awesome, but with what dependency and tools versions?

**Comment by Tyler-V** [comment]
@giggio 

I'm using all @latest packages with the proper scoped lib configurations as suggested by @filipesilva 

You can use my project as a reference on how I am serving my demo to develop the libraries with refresh.

**Comment by banminkyoz** [comment]
@giggio I also use that way :cat:. With projects do have UI. It's really paintful to rebuild library and re-run everytime we do have a change :cry:

**Comment by giggio** [comment]
@hansl how is this going to work and in what version of the CLI should we expect it? Is it 6.1.0 as per the milestone?

**Comment by alan-agius4** [comment]
@giggio, you can read the documentation here: https://github.com/alan-agius4/angular-cli/blob/master/docs/documentation/stories/create-library.md#why-do-i-need-to-build-the-library-everytime-i-make-changes

In a nutshell you'd need to run
```bash
ng build my-lib --watch
```

As far as the release, this won't be under 6.1.x.

**Comment by GeorgeKnap** [comment]
I've been using the `ng build my-lib --watch` for some time now but I 've been unhappy  with the rebuild time.
Today I changed the default path mapping in main project's tsconfig to point to lib's public_api:

```
"paths": {
      "users": [
        "projects/users/src/public_api"
      ],
      "shared": [
        "projects/shared/src/public_api"
      ]
    }
```
now the libraries are built by `ng serve` command, changes propagates super fast and also the Chrome devtools are able to debug raw typescript files.
Sounds too good to be true but really works.

**Comment by playground** [comment]
@filipesilva how would you handle project with multiple library dependencies?
For example project A depends on lib-B and lib-C, if changes are made to lib-C, will building project A only compile lib-C?

```
Heya, this feature request is something we are aware of and talk about in https://github.com/angular/angular-cli/wiki/stories-create-library#why-do-i-need-to-build-the-library-everytime-i-make-changes.
```

**Comment by Noah1989** [comment]
On my machine rebuilding the library in a separate process triggers the rebuild of the main app too early and the resulting library code in the app is always one change behind. The solution by @GeorgeKnap works, though.

**Comment by vimalans** [comment]
@GeorgeKnap 's suggestion works!

**Comment by nickofthyme** [comment]
@GeorgeKnap you're a lifesaver! Thanks 🎉

**Comment by rdhelms** [comment]
@GeorgeKnap 

Wow...so that really does work eerily well...do you know what the downsides are of using the `projects` path instead of `dist`? Surely there must be some reasons NOT to do that or else angular would have it be the default / recommended path?

**Comment by GeorgeKnap** [comment]
@rdhelms 
The point is you are pointing to the library source, not the library output folder. That way `ng serve` works with source typescript files.
However I think this is good only for development. For prod environment I have the libraries built separately in npm feed and using them as any other 3rd party dependency.

Does it answer your question?

**Comment by giggio** [comment]
I came back to this, as it has been a while since I tried the whole "double watch" solution, and it could have improved on ng7.
Not at all. It is still much slower than running everything as a single project.
I have just updated my whole project and deps to version 7 and I can say it is better than before, but still to slow.
I hope a solution eventually comes up, until then I will keep using the tsconfig path remap.
